src/pm/xbps.rs

```diff
@@ -0,0 +1,278 @@
+                .await;
+        }
+
+        let lines: Vec<_> = stream::iter(kws)
+            .map(Ok)
+            .and_then(|&pkg| async {
+                let cmd = Cmd::new(&["xbps-query", "--property", "pkgver", pkg]).flags(flags);
+                self.check_output(cmd, PmMode::Mute, &Strategy::default())
+                    .await
+                    .map_err(|err| (pkg.to_string(), err))
+            })
+            .collect()
+            .await;
+
+        let mut stdout = tokio::io::stdout();
+        let mut pkg_not_found_detected = false;
+        for line in lines {
+            match line {
+                Ok(line) => {
+                    stdout.write_all(&line).await?;
+                }
+                Err((
+                    pkg,
+                    err @ Error::CmdStatusCodeError {
+                        code: PKG_NOT_FOUND_CODE,
+                        ..
+                    },
+                )) => {
+                    pkg_not_found_detected = true;
+                    let msg = format!("package '{}' was not found", pkg);
+                    print_err(err, &msg);
+                }
+                Err((_, other)) => return Err(other),
+            };
+        }
+
+        if pkg_not_found_detected {
+            return Err(Error::CmdStatusCodeError {
+                code: PKG_NOT_FOUND_CODE,
+                output: vec![],
+            });
+        }
+
+        Ok(())
+    }
+
+    /// Qe lists packages installed explicitly (not as dependencies).
+    async fn qe(&self, kws: &[&str], flags: &[&str]) -> Result<()> {
+        if kws.is_empty() {
+            return self
+                .run(Cmd::new(&["xbps-query", "-m"]).kws(kws).flags(flags))
+                .await;
+        }
+
+        let lines: Vec<_> = stream::iter(kws)
+            .filter(|&&pkg| async {
+                let check_cmd =
+                    Cmd::new(&["xbps-query", "--property", "automatic-install", pkg]).flags(flags);
+                let result = self
+                    .check_output(check_cmd, PmMode::Mute, &Strategy::default())
+                    .await;
+
+                // if a package is manually installed then the automatic-install field is empty
+                match result {
+                    Ok(auto_status) => auto_status.is_empty(),
+                    _ => true,
+                }
+            })
+            .map(Ok)
+            .and_then(|&pkg| async {
+                let cmd = Cmd::new(&["xbps-query", "--property", "pkgver", pkg]).flags(flags);
+                self.check_output(cmd, PmMode::Mute, &Strategy::default())
+                    .await
+                    .map_err(|err| (pkg.to_string(), err))
+            })
+            .collect()
+            .await;
+
+        let mut stdout = tokio::io::stdout();
+        let mut pkg_not_found_detected = false;
+        for line in lines {
+            match line {
+                Ok(line) => {
+                    stdout.write_all(&line).await?;
```

Comment:
Using `.write_all` is dangerous here, as this operation should be blocking.
In fact, https://docs.rs/tokio/latest/tokio/io/struct.Stdout.html says:

> In particular you should be aware that writes using [write_all](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.write_all) are not guaranteed to occur as a single write, so multiple threads writing data with [write_all](https://docs.rs/tokio/latest/tokio/io/trait.AsyncWriteExt.html#method.write_all) may result in interleaved output.

Reply:
I guess we don't gain anything here from using `tokio::io::stdout()` over `std::io::stdout()` which is globally mutexed instead, so I'll switch to that